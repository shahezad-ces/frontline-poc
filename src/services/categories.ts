import {
  GET_CATEGORIES_QUERY,
  GET_CATEGORY_DETAILS_QUERY,
} from "@frontline/graphql";
import { query } from "@frontline/libs/ApolloClient";
import { Category } from "@frontline/types";

export const getCategories = async (): Promise<Category[] | undefined> => {
  try {
    const { data } = await query<{ categories: Category[] }>({
      query: GET_CATEGORIES_QUERY,
      fetchPolicy: "cache-first",
    });

    return data?.categories;
  } catch (error) {
    console.error("GraphQL Error in getCategories:", error);
    return undefined;
  }
};

export const getCategoryDetails = async (
  id: number
): Promise<Category | undefined> => {
  if (!id) return undefined;
  try {
    const { data } = await query<{ category: Category }>({
      query: GET_CATEGORY_DETAILS_QUERY,
      variables: { id },
      fetchPolicy: "cache-first",
    });

    return data?.category;
  } catch (error) {
    console.error("GraphQL Error in getCategoryDetails:", error);
    return undefined;
  }
};
