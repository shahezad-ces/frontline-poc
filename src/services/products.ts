import {
  GET_PRODUCTS_QUERY,
  GET_PRODUCT_DETAILS_QUERY,
} from "@frontline/graphql";
import { query } from "@frontline/libs/ApolloClient";
import type { Product } from "@frontline/types";

export interface GetProductsParams {
  categoryId?: number;
  price?: number;
  price_min?: number;
  price_max?: number;
  offset?: number;
  limit?: number;
}

export const getProducts = async (
  params: GetProductsParams
): Promise<Product[] | undefined> => {
  const {
    categoryId,
    price,
    price_min,
    price_max,
    offset = 0,
    limit = 20,
  } = params;
  const variables = {
    offset,
    limit,
    ...(categoryId !== undefined && { categoryId }),
    ...(price !== undefined && { price }),
    ...(price_min !== undefined && { price_min }),
    ...(price_max !== undefined && { price_max }),
  };

  const { data } = await query<{ products: Product[] }>({
    query: GET_PRODUCTS_QUERY,
    variables,
    fetchPolicy: "cache-first",
  });

  return data?.products;
};

export const getProductDetails = async (
  productId: number
): Promise<Product | undefined> => {
  try {
    const { data } = await query<{ product: Product }>({
      query: GET_PRODUCT_DETAILS_QUERY,
      variables: { id: productId },
      fetchPolicy: "cache-first",
    });

    return data?.product;
  } catch (error) {
    console.error("GraphQL Error:", error);
    return undefined; // fallback
  }
};
