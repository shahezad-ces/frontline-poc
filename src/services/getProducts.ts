import { GET_PRODUCTS_QUERY } from "@frontline/graphql/products";
import { fetchGraphQL } from "@frontline/libs/apolloClient";
import { Product } from "@frontline/types/product";

export async function getProducts(): Promise<Product[]> {
  const variables = {
    offset: 0,
    limit: 20,
  };

  const data = await fetchGraphQL<{ products: Product[] }>({
    query: GET_PRODUCTS_QUERY,
    variables,
  });

  return data.products;
}
