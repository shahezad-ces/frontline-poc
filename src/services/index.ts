import { getCategories, getCategoryDetails } from "./categories";
import { getProducts, getProductDetails } from "./products";

export { getCategories, getCategoryDetails, getProducts, getProductDetails };
