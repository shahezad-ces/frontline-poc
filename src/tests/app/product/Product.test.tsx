import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import ProductDetailPage from "@frontline/app/product/[id]/page";
import { getProductDetails } from "@frontline/services";

// Mock services
jest.mock("@frontline/services", () => ({
  getProductDetails: jest.fn(),
}));

// Mock child components to avoid client-side hook issues
jest.mock("@frontline/components", () => ({
  AddToCartButton: ({ product }: any) => (
    <button data-testid="add-to-cart">Add to Cart</button>
  ),
}));

jest.mock("@frontline/stories", () => ({
  Image: ({ src, alt }: any) => (
    <img src={src} alt={alt} data-testid="product-image" />
  ),
  QuantitySelector: () => (
    <div data-testid="quantity-selector">Quantity Selector</div>
  ),
}));

const mockGetProductDetails = getProductDetails as jest.Mock;

const mockProduct = {
  id: 1,
  title: "Test Product",
  price: 199.99,
  images: ["https://placehold.co/600x400"],
  category: { id: 1, name: "Electronics", image: "cat.jpg" },
  description: "A great product for testing.",
};

describe("ProductDetailPage (Server Component)", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders product details when data is available", async () => {
    mockGetProductDetails.mockResolvedValue(mockProduct);

    const ui = await ProductDetailPage({ params: Promise.resolve({ id: 1 }) });
    render(ui);

    expect(screen.getByText(/Test Product/i)).toBeInTheDocument();
    expect(screen.getByText(/Electronics/i)).toBeInTheDocument();
    expect(screen.getByText(/\$199.99/i)).toBeInTheDocument();
    expect(screen.getByTestId("product-image")).toHaveAttribute(
      "src",
      mockProduct.images[0]
    );
    expect(screen.getByTestId("quantity-selector")).toBeInTheDocument();
    expect(screen.getByTestId("add-to-cart")).toBeInTheDocument();
  });

  it("displays error message when product details are missing", async () => {
    mockGetProductDetails.mockResolvedValue(undefined);

    const ui = await ProductDetailPage({ params: Promise.resolve({ id: 1 }) });
    render(ui);

    expect(
      screen.getByText(
        /Failed to load product details. Please try again later./i
      )
    ).toBeInTheDocument();
  });

  it("matches snapshot", async () => {
    mockGetProductDetails.mockResolvedValue(mockProduct);

    const ui = await ProductDetailPage({ params: Promise.resolve({ id: 1 }) });
    const { asFragment } = render(ui);
    expect(asFragment()).toMatchSnapshot();
  });
});
