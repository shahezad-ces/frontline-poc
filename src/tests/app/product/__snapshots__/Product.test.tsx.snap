// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProductDetailPage (Server Component) matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto my-6 px-6 py-8 grid grid-cols-1 lg:grid-cols-12 gap-8 bg-white border border-gray-200 rounded-lg"
  >
    <div
      class="lg:col-span-4"
    >
      <div
        class="border border-gray-400 rounded-lg overflow-hidden"
      >
        <img
          alt="Test Product"
          data-testid="product-image"
          src="https://placehold.co/600x400"
        />
      </div>
    </div>
    <div
      class="lg:col-span-8 space-y-2"
    >
      <h1
        class="text-3xl font-bold"
      >
        Test Product
      </h1>
      <p
        class="text-gray-500 text-sm"
      >
        Electronics
      </p>
      <p
        class="text-red-600 text-3xl font-bold"
      >
        $199.99 
      </p>
      <p
        class="text-green-600 font-semibold"
      >
        Save $350 (26%)
      </p>
      <div
        data-testid="quantity-selector"
      >
        Quantity Selector
      </div>
      <p
        class="text-gray-500 text-sm mt-6"
      >
        A great product for testing.
      </p>
      <button
        data-testid="add-to-cart"
      >
        Add to Cart
      </button>
    </div>
  </div>
</DocumentFragment>
`;
